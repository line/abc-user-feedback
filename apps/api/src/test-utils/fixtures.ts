/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import * as bcrypt from 'bcrypt';

import {
  FieldFormatEnum,
  FieldStatusEnum,
  FieldTypeEnum,
  isSelectFieldFormat,
} from '@/common/enums';
import type { ChannelEntity } from '@/domains/admin/channel/channel/channel.entity';
import type {
  CreateFieldDto,
  ReplaceFieldDto,
} from '@/domains/admin/channel/field/dtos';
import type { FieldEntity } from '@/domains/admin/channel/field/field.entity';
import type { FeedbackEntity } from '@/domains/admin/feedback/feedback.entity';
import type { CreateIssueDto } from '@/domains/admin/project/issue/dtos';
import type { TenantEntity } from '@/domains/admin/tenant/tenant.entity';
import {
  SignUpMethodEnum,
  UserStateEnum,
  UserTypeEnum,
} from '@/domains/admin/user/entities/enums';
import type { UserEntity } from '@/domains/admin/user/entities/user.entity';

export const createFieldEntity = (input: Partial<CreateFieldDto>) => {
  const format = input?.format ?? getRandomEnumValue(FieldFormatEnum);
  const type = input?.type ?? getRandomEnumValue(FieldTypeEnum);
  const status = input?.status ?? getRandomEnumValue(FieldStatusEnum);
  return {
    name: faker.string.alphanumeric(20),
    description: faker.lorem.lines(2),
    format,
    type,
    status,
    options:
      format === FieldFormatEnum.select
        ? getRandomOptionEntities().sort(optionSort)
        : undefined,
    ...input,
  };
};
export const createFieldDto = (input: Partial<CreateFieldDto>) => {
  const format = input?.format ?? getRandomEnumValue(FieldFormatEnum);
  const type = input?.type ?? getRandomEnumValue(FieldTypeEnum);
  const status = input?.status ?? getRandomEnumValue(FieldStatusEnum);
  return {
    name: faker.string.alphanumeric(20),
    key: faker.string.alphanumeric(20),
    description: faker.lorem.lines(2),
    format,
    type,
    status,
    options: isSelectFieldFormat(format)
      ? getRandomOptionDtos().sort(optionSort)
      : undefined,
    ...input,
  };
};
export const updateFieldDto = (input: Partial<ReplaceFieldDto>) => {
  return {
    id: faker.number.int(),
    ...createFieldDto(input),
  };
};

export const createIssueDto = (input: Partial<CreateIssueDto>) => {
  return {
    name: faker.string.alphanumeric(20),
    ...input,
  };
};

export const getRandomValue = (
  format: FieldFormatEnum,
  options?: { id: number; name: string; key: string }[],
) => {
  switch (format) {
    case FieldFormatEnum.text:
      return faker.string.sample();
    case FieldFormatEnum.keyword:
      return faker.string.sample();
    case FieldFormatEnum.number:
      return faker.number.int();
    case FieldFormatEnum.boolean:
      return faker.datatype.boolean();
    case FieldFormatEnum.select:
      return options.length === 0
        ? undefined
        : options[faker.number.int({ min: 0, max: options.length - 1 })].key;
    case FieldFormatEnum.multiSelect:
      return options.length === 0
        ? []
        : faker.helpers
            .shuffle(options)
            .slice(0, faker.number.int({ min: 0, max: options.length - 1 }))
            .map((option) => option.key);
    case FieldFormatEnum.date:
      return faker.date.anytime().toISOString();
    case FieldFormatEnum.image:
      return faker.string.sample();
    default:
      throw new Error('Invalid field type ');
  }
};

const getRandomOptionEntities = () => {
  const length = faker.number.int(10);
  return Array.from({ length }).map(() => ({
    id: faker.number.int(),
    name: faker.string.sample(),
  }));
};
const getRandomOptionDtos = () => {
  const length = faker.number.int(10);
  return Array.from({ length }).map(() => {
    const randomValue = faker.string.sample();
    return {
      id: faker.number.int(),
      name: randomValue,
      key: randomValue,
    };
  });
};

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.number.int(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const optionSort = (a, b) =>
  a.name < b.name ? -1 : a.name > b.name ? 1 : 0;

export const fieldsFixture = Object.values(FieldFormatEnum).flatMap((format) =>
  Object.values(FieldTypeEnum).flatMap((type) =>
    Object.values(FieldStatusEnum).flatMap((status) => ({
      id: faker.number.int(),
      ...createFieldDto({
        format,
        type,
        status,
      }),
    })),
  ),
) as FieldEntity[];

export const feedbackDataFixture = fieldsFixture.reduce((prev, curr) => {
  if (curr.type === FieldTypeEnum.ADMIN) return prev;
  if (curr.status === FieldStatusEnum.INACTIVE) return prev;
  const value = getRandomValue(curr.format, curr.options);
  return {
    ...prev,
    [curr.key]: value,
  };
}, {});

export const feedbackFixture = {
  id: faker.number.int(),
  rawData: feedbackDataFixture,
  additionalData: {},
  createdAt: faker.date.past(),
  updatedAt: faker.date.past(),
  channel: {
    id: faker.number.int(),
    name: faker.string.sample(),
    description: faker.lorem.lines(2),
    imageConfig: null,
    createdAt: faker.date.past(),
    updatedAt: faker.date.past(),
  } as ChannelEntity,
  issues: [],
} as FeedbackEntity;

export const passwordFixture = faker.internet.password();

export const emailFixture = faker.internet.email();

export const userFixture = {
  id: faker.number.int(),
  email: emailFixture,
  name: faker.string.sample(),
  department: faker.string.sample(),
  state: getRandomEnumValue(UserStateEnum),
  hashPassword: bcrypt.hashSync(passwordFixture, 0),
  type: getRandomEnumValue(UserTypeEnum),
  signUpMethod: getRandomEnumValue(SignUpMethodEnum),
} as UserEntity;

export const tenantFixture = {
  id: faker.number.int(),
  siteName: faker.string.sample(),
  description: faker.lorem.lines(2),
  useEmail: faker.datatype.boolean(),
  isPrivate: faker.datatype.boolean(),
  isRestrictDomain: faker.datatype.boolean(),
  allowDomains: [],
  useOAuth: faker.datatype.boolean(),
  oauthConfig: null,
  createdAt: faker.date.past(),
  updatedAt: faker.date.past(),
} as TenantEntity;
