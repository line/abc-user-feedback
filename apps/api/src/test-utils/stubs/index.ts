/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
export { ApiKeyRepositoryStub } from './api-key-repository.stub';
export { ChannelRepositoryStub } from './channel-repository.stub';
export { CodeRepositoryStub } from './code-repository.stub';
export { EventRepositoryStub } from './event-repository.stub';
export { FeedbackRepositoryStub } from './feedback-repository.stub';
export { FieldRepositoryStub } from './field-repository.stub';
export { IssueRepositoryStub } from './issue-repository.stub';
export { ProjectRepositoryStub } from './project-repository.stub';
export { RoleRepositoryStub } from './role-repository.stub';
export { TenantRepositoryStub } from './tenant-repository.stub';
export { UserRepositoryStub } from './user-repository.stub';
export { WebhookRepositoryStub } from './webhook-repository.stub';
