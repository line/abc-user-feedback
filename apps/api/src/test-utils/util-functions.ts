/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import type { InjectionToken, Provider } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import type { DataSource, Repository } from 'typeorm';
import { initializeTransactionalContext } from 'typeorm-transactional';

import { jwtConfig, jwtConfigSchema } from '@/configs/jwt.config';
import {
  opensearchConfig,
  opensearchConfigSchema,
} from '@/configs/opensearch.config';
import { smtpConfig, smtpConfigSchema } from '@/configs/smtp.config';
import type { AuthService } from '@/domains/admin/auth/auth.service';
import { UserDto } from '@/domains/admin/user/dtos';
import { UserStateEnum } from '@/domains/admin/user/entities/enums';
import { UserEntity } from '@/domains/admin/user/entities/user.entity';

initializeTransactionalContext();

export const getMockProvider = (
  injectToken: InjectionToken,
  factory: unknown,
): Provider => ({ provide: injectToken, useFactory: () => factory });

export const TestConfig = ConfigModule.forRoot({
  load: [smtpConfig, jwtConfig, opensearchConfig],
  envFilePath: '.env.test',
  validationSchema: smtpConfigSchema
    .concat(jwtConfigSchema)
    .concat(opensearchConfigSchema),
});

export const MockDataSource = {
  initialize: jest.fn(),
};

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.number.int(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const clearEntities = async (repos: Repository<any>[]) => {
  for (const repo of repos) {
    await repo.query('set foreign_key_checks = 0');
    await repo.clear();
    await repo.query('set foreign_key_checks = 1');
  }
};

export const signInTestUser = async (
  dataSource: DataSource,
  authService: AuthService,
) => {
  const userRepo = dataSource.getRepository(UserEntity);
  const user = await userRepo.save({
    email: faker.internet.email(),
    state: UserStateEnum.Active,
    hashPassword: faker.internet.password(),
  });
  return { jwt: await authService.signIn(UserDto.transform(user)), user };
};

export const DEFAULT_FIELD_COUNT = 2;

export const createQueryBuilder: any = {
  setFindOptions: () => jest.fn().mockImplementation(() => createQueryBuilder),
  select: () => createQueryBuilder,
  innerJoin: () => createQueryBuilder,
  leftJoin: () => createQueryBuilder,
  where: () => createQueryBuilder,
  andWhere: () => createQueryBuilder,
  groupBy: () => createQueryBuilder,
  addOrderBy: () => createQueryBuilder,
  getRawMany: () => createQueryBuilder,
  insert: () => createQueryBuilder,
  values: () => createQueryBuilder,
  orUpdate: () => createQueryBuilder,
  updateEntity: () => createQueryBuilder,
  execute: () => createQueryBuilder,
  offset: () => createQueryBuilder,
  limit: () => createQueryBuilder,
  getMany: () => createQueryBuilder,
  getCount: () => createQueryBuilder,
};

export const mockRepository = () => ({
  findOneBy: jest.fn(),
  findOne: jest.fn(),
  find: jest.fn(),
  findBy: jest.fn(),
  findAndCount: jest.fn(),
  findAndCountBy: jest.fn(),
  save: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  upsert: jest.fn(),
  count: jest.fn(),
  remove: jest.fn(),
  createQueryBuilder: jest.fn(() => createQueryBuilder),
  query: jest.fn(),
});

export const MockOpensearchRepository = {
  createIndex: jest.fn(),
  deleteIndex: jest.fn(),
  putMappings: jest.fn(),
  createData: jest.fn(),
  getData: jest.fn(),
  updateData: jest.fn(),
  getTotal: jest.fn(),
};
