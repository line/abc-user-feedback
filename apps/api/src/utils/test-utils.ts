/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import { MailerModule } from '@nestjs-modules/mailer';
import { InjectionToken, Provider } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { TypeOrmModule } from '@nestjs/typeorm';
import { join } from 'path';
import { DataSource, Repository } from 'typeorm';
import { SnakeNamingStrategy } from 'typeorm-naming-strategies';
import {
  addTransactionalDataSource,
  initializeTransactionalContext,
} from 'typeorm-transactional';

import { jwtConfig, jwtConfigSchema } from '@/configs/jwt.config';
import { mySqlConfigSchema, mysqlConfig } from '@/configs/mysql.config';
import { smtpConfig, smtpConfigSchema } from '@/configs/smtp.config';
import { AuthService } from '@/domains/auth/auth.service';
import { UserDto } from '@/domains/user/dtos';
import { UserStateEnum } from '@/domains/user/entities/enums';
import { UserEntity } from '@/domains/user/entities/user.entity';
import { ConfigServiceType } from '@/types/config-service.type';

initializeTransactionalContext();

export const getMockProvider = (
  injectToken: InjectionToken,
  factory: unknown,
): Provider => ({ provide: injectToken, useFactory: () => factory });

export const TestConfig = ConfigModule.forRoot({
  load: [smtpConfig, jwtConfig, mysqlConfig],
  envFilePath: '.env.test',
  validate: (config) => ({
    ...smtpConfigSchema.validateSync(config),
    ...jwtConfigSchema.validateSync(config),
    ...mySqlConfigSchema.validateSync(config),
  }),
});

export const TestMailConfigModule = MailerModule.forRoot({
  transport: { url: '' },
});

export const TestTypeOrmConfig = TypeOrmModule.forRootAsync({
  imports: [ConfigModule],
  inject: [ConfigService],
  dataSourceFactory: async (options) => {
    const datasource = await new DataSource(options).initialize();
    return addTransactionalDataSource(datasource);
  },
  useFactory: (configService: ConfigService<ConfigServiceType>) => {
    const { main_url, sub_urls } = configService.get('mysql', {
      infer: true,
    });
    return {
      type: 'mysql',
      replication: {
        master: { url: main_url },
        slaves: sub_urls.map((url) => ({ url })),
      },
      entities: [join(__dirname, '../**/*.entity.{ts,js}')],
      logging: ['warn', 'error'],
      namingStrategy: new SnakeNamingStrategy(),
      timezone: '+00:00',
    };
  },
});

export const TestConfigs = [
  TestConfig,
  TestMailConfigModule,
  TestTypeOrmConfig,
];

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.datatype.number(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const clearEntities = async (repos: Repository<any>[]) => {
  for (const repo of repos) {
    await repo.query('set foreign_key_checks = 0');
    await repo.clear();
    await repo.query('set foreign_key_checks = 1');
  }
};

export const signInTestUser = async (
  dataSource: DataSource,
  authService: AuthService,
) => {
  const userRepo = dataSource.getRepository(UserEntity);
  const user = await userRepo.save({
    email: faker.internet.email(),
    state: UserStateEnum.Active,
    hashPassword: faker.internet.password(),
  });
  return { jwt: await authService.signIn(UserDto.transform(user)), user };
};

export const DEFAULT_FIELD_COUNT = 2;

export const createQueryBuilder: any = {
  setFindOptions: () =>
    jest.fn().mockImplementation(() => {
      return createQueryBuilder;
    }),
};

export const mockRepository = () => ({
  findOneBy: jest.fn(),
  findOne: jest.fn(),
  find: jest.fn(),
  findBy: jest.fn(),
  findAndCount: jest.fn(),
  findAndCountBy: jest.fn(),
  save: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  upsert: jest.fn(),
  count: jest.fn(),
  remove: jest.fn(),
  createQueryBuilder: jest.fn(() => createQueryBuilder),
  query: jest.fn(),
});

export const MockOpensearchRepository = {
  createIndex: jest.fn(),
  deleteIndex: jest.fn(),
  putMappings: jest.fn(),
  createData: jest.fn(),
  getData: jest.fn(),
  updateData: jest.fn(),
  getTotal: jest.fn(),
};
