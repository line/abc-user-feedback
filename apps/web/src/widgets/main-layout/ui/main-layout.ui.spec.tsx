/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import mockRouter from 'next-router-mock';

import { render, screen } from '@/test-utils';
import MainLayout from './main-layout.ui';

describe('MainLayout', () => {
  beforeEach(() => {
    mockRouter.pathname = '/';
  });
  test('snapshot default', () => {
    const { container } = render(<MainLayout />);
    expect(container).toMatchSnapshot();
  });

  test('snapshot center', () => {
    const { container } = render(<MainLayout center />);
    expect(container).toMatchSnapshot();
  });

  test('snapshot with sidenav', () => {
    mockRouter.pathname = '/main/project/[projectId]';
    const { container } = render(<MainLayout />);

    expect(container).toMatchSnapshot();
  });

  test('should render children', () => {
    const TEXT = faker.string.sample();
    render(<MainLayout>{TEXT}</MainLayout>);
    expect(screen.getByText(TEXT)).toBeInTheDocument();
  });
});
