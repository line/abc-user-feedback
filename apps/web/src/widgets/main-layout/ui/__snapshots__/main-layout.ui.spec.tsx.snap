// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout snapshot center 1`] = `
<div>
  <div
    class="min-w-[960px]"
  >
    <header
      class="relative flex h-[48px] items-center justify-between px-4"
    >
      <div
        class="flex flex-1 items-center gap-6"
      >
        <a
          class="flex cursor-pointer items-center gap-1"
          href="/auth/sign-in"
        >
          <img
            alt="logo"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/assets/images/logo.svg"
            style="color: transparent;"
            width="24"
          />
          <svg
            class="inline-block h-[24px] w-[123px]"
            height="24"
            width="24"
          />
        </a>
        <div
          class="flex items-center gap-2"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="bg-fill-tertiary inline-flex rounded-sm p-1.5"
            >
              <svg
                class="inline-block text-secondary"
                height="12"
                width="12"
              />
            </div>
            <span
              class="font-12-regular"
            />
          </div>
        </div>
      </div>
      <div
        class="flex items-center gap-2 self-stretch"
      >
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="btn btn-sm btn-secondary min-w-0 px-2"
            data-headlessui-state=""
            id="headlessui-listbox-button-:r8:"
            type="button"
          >
            <svg
              class="inline-block mr-1"
              height="16"
              width="16"
            />
            <span
              class="font-12-bold uppercase"
            />
          </button>
        </div>
        <button
          class="icon-btn icon-btn-sm icon-btn-secondary"
        >
          <svg
            class="inline-block text-primary"
            height="24"
            width="24"
          />
        </button>
      </div>
    </header>
    <div
      class="bg-primary border-fill-secondary absolute left-1/2 top-1/2 min-w-[440px] -translate-x-1/2 -translate-y-1/2 rounded border p-8"
    />
    <footer
      class="font-14-bold text-secondary absolute bottom-[7%] left-1/2 z-10 -translate-x-1/2"
    >
      Â© ABC Studio All rights reserved
    </footer>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`MainLayout snapshot default 1`] = `
<div>
  <div
    class="min-w-[960px]"
  >
    <header
      class="relative flex h-[48px] items-center justify-between px-4"
    >
      <div
        class="flex flex-1 items-center gap-6"
      >
        <a
          class="flex cursor-pointer items-center gap-1"
          href="/auth/sign-in"
        >
          <img
            alt="logo"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/assets/images/logo.svg"
            style="color: transparent;"
            width="24"
          />
          <svg
            class="inline-block h-[24px] w-[123px]"
            height="24"
            width="24"
          />
        </a>
        <div
          class="flex items-center gap-2"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="bg-fill-tertiary inline-flex rounded-sm p-1.5"
            >
              <svg
                class="inline-block text-secondary"
                height="12"
                width="12"
              />
            </div>
            <span
              class="font-12-regular"
            />
          </div>
        </div>
      </div>
      <div
        class="flex items-center gap-2 self-stretch"
      >
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="btn btn-sm btn-secondary min-w-0 px-2"
            data-headlessui-state=""
            id="headlessui-listbox-button-:r2:"
            type="button"
          >
            <svg
              class="inline-block mr-1"
              height="16"
              width="16"
            />
            <span
              class="font-12-bold uppercase"
            />
          </button>
        </div>
        <button
          class="icon-btn icon-btn-sm icon-btn-secondary"
        >
          <svg
            class="inline-block text-primary"
            height="24"
            width="24"
          />
        </button>
      </div>
    </header>
    <div
      class="flex min-h-[calc(100vh-48px)] items-stretch"
    >
      <main
        class="mx-4 my-6 w-full overflow-x-auto"
      />
    </div>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;

exports[`MainLayout snapshot with sidenav 1`] = `
<div>
  <div
    class="min-w-[960px]"
  >
    <header
      class="relative flex h-[48px] items-center justify-between px-4"
    >
      <div
        class="flex flex-1 items-center gap-6"
      >
        <a
          class="flex cursor-pointer items-center gap-1"
          href="/main"
        >
          <img
            alt="logo"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/assets/images/logo.svg"
            style="color: transparent;"
            width="24"
          />
          <svg
            class="inline-block h-[24px] w-[123px]"
            height="24"
            width="24"
          />
        </a>
        <div
          class="flex items-center gap-2"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="bg-fill-tertiary inline-flex rounded-sm p-1.5"
            >
              <svg
                class="inline-block text-secondary"
                height="12"
                width="12"
              />
            </div>
            <span
              class="font-12-regular"
            />
          </div>
        </div>
      </div>
      <div
        class="flex items-center gap-2 self-stretch"
      >
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            class="btn btn-sm btn-secondary min-w-0 px-2"
            data-headlessui-state=""
            id="headlessui-listbox-button-:re:"
            type="button"
          >
            <svg
              class="inline-block mr-1"
              height="16"
              width="16"
            />
            <span
              class="font-12-bold uppercase"
            />
          </button>
        </div>
        <button
          class="icon-btn icon-btn-sm icon-btn-secondary"
        >
          <svg
            class="inline-block text-primary"
            height="24"
            width="24"
          />
        </button>
      </div>
    </header>
    <div
      class="flex min-h-[calc(100vh-48px)] items-stretch"
    >
      <nav
        class="relative"
      >
        <div
          class="h-full w-[72px]"
        />
        <ul
          class="absolute left-0 top-0 z-20 h-full space-y-1 p-4"
        >
          <li>
            <a
              href="/main/project//dashboard"
            >
              <button
                class="icon-btn icon-btn-tertiary icon-btn-md w-full flex-nowrap justify-start"
              >
                <svg
                  class="inline-block "
                  height="24"
                  width="24"
                />
                <span
                  class="ml-2 hidden"
                >
                  main.dashboard.title
                </span>
              </button>
            </a>
          </li>
          <hr />
          <li>
            <a
              href="/main/project//feedback"
            >
              <button
                class="icon-btn icon-btn-tertiary icon-btn-md w-full flex-nowrap justify-start"
              >
                <svg
                  class="inline-block "
                  height="24"
                  width="24"
                />
                <span
                  class="ml-2 hidden"
                >
                  main.feedback.title
                </span>
              </button>
            </a>
          </li>
          <li>
            <a
              href="/main/project//issue"
            >
              <button
                class="icon-btn icon-btn-tertiary icon-btn-md w-full flex-nowrap justify-start"
              >
                <svg
                  class="inline-block "
                  height="24"
                  width="24"
                />
                <span
                  class="ml-2 hidden"
                >
                  main.issue.title
                </span>
              </button>
            </a>
          </li>
          <hr />
          <li>
            <a
              href="/main/project//setting"
            >
              <button
                class="icon-btn icon-btn-tertiary icon-btn-md w-full flex-nowrap justify-start"
              >
                <svg
                  class="inline-block "
                  height="24"
                  width="24"
                />
                <span
                  class="ml-2 hidden"
                >
                  main.setting.title
                </span>
              </button>
            </a>
          </li>
        </ul>
      </nav>
      <main
        class="mx-4 my-6 w-full overflow-x-auto"
      />
    </div>
  </div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;
